### PR TITLE
Add Provisioning token status creation/update functionality

### PR DESCRIPTION
**Pull request by fabribet**
### Description

Adds functionality to test the newly introduced WEBHOOK `CARD_TOKEN_LIFECYCLE`. Allows to create/update a Provisioning token by triggering the webhook.

### Screenshots
Screen Shot 2020-06-18 at 2 40 29 PM

### CODE REVIEW

File: src/helpers/cards.ts [hunk 1]
```diff
@@ -321,6 +323,165 @@ export const changeCardStatus = async (
...
+    throw new Error("You have to provide cardId");
+  }
+
+  const person = await db.getPerson(personId);
+  const cardData = person.account.cards.find(({ card }) => card.id === cardId);
+  if (!cardData) {
+    throw new Error("Card not found");
+  }
+
+  const { provisioningToken } = cardData;
```
Comment (p-janik) [inline src/helpers/cards.ts]: up to this point in his function, it's the same logic as in `createProvisioningToken`, maybe we could extract and reuse it?

File: src/helpers/cards.ts [hunk 2]
```diff
@@ -321,6 +323,165 @@ export const changeCardStatus = async (
...
+    token_reference_id: uuid.v4(),
+    client_wallet_account_id: walletId || uuid.v4(),
+    wallet_type: "GOOGLE",
+  };
+
+  // Token Creation
+  let payload: ProvisioningTokenStatusChangePayload = {
+    ...baseData,
+    token_status: ProvisioningTokenStatus.INACTIVE,
+    event_type: "TOKEN_CREATED",
```
Comment (p-janik) [inline src/helpers/cards.ts]: you could extract these to enum

File: src/helpers/types.ts [hunk 3]
```diff
@@ -186,7 +203,8 @@ export enum CardWebhookEvent {
   "CARD_AUTHORIZATION_DECLINE" = "CARD_AUTHORIZATION_DECLINE", // 	The transaction was declined.
   "CARD_AUTHORIZATION_RESOLUTION" = "CARD_AUTHORIZATION_RESOLUTION", // 	The reservation was cancelled, expired or booked. No push notification for the customer is required for this webhook.
   "CARD_LIFECYCLE_EVENT" = "CARD_LIFECYCLE_EVENT", // The status of the card is changed.
-  "CARD_TOKEN_LIFECYCLE" = "CARD_TOKEN_LIFECYCLE",
+  "CARD_TOKEN_LIFECYCLE" = "CARD_TOKEN_LIFECYCLE", // Push provisioning token is created or updated.
+  "CARD_FRAUD_CASE" = "CARD_FRAUD_CASE",
```
Comment (p-janik) [inline src/helpers/types.ts]: I'm not sure about this one, we already have `pending` and `timeout` events

Comment (fabribet) [reply src/helpers/types.ts]: 🤦 it was a merging conflict that I solved incorrectly

File: src/helpers/cards.ts [hunk 4]
```diff
@@ -321,6 +323,165 @@ export const changeCardStatus = async (
...
+  if (!cardData) {
+    throw new Error("Card not found");
+  }
+
+  const { provisioningToken } = cardData;
+  let walletId;
+
+  // Deactivate existing token
+  if (provisioningToken) {
+    walletId = provisioningToken.client_wallet_account_id
```
Comment (teodorpatras) [inline src/helpers/cards.ts]: nit: missing trailing `;`

File: src/helpers/cards.ts [hunk 5]
```diff
@@ -321,6 +323,165 @@ export const changeCardStatus = async (
...
+    const data = {
+      card_id: cardData.card.id,
+      token_status: ProvisioningTokenStatus.DEACTIVATED,
+      token_reference_id: provisioningToken.token_reference_id,
+      client_wallet_account_id: walletId,
+      wallet_type: "GOOGLE",
```
Comment (teodorpatras) [inline src/helpers/cards.ts]: could these all strings be enums?

File: src/helpers/cards.ts [hunk 6]
```diff
@@ -321,6 +323,165 @@ export const changeCardStatus = async (
...
+  };
+
+  // Token Creation
+  let payload: ProvisioningTokenStatusChangePayload = {
+    ...baseData,
+    token_status: ProvisioningTokenStatus.INACTIVE,
+    event_type: "TOKEN_CREATED",
+    message_reason: "TOKEN_CREATED",
+  }
+  await triggerWebhook(CardWebhookEvent.CARD_TOKEN_LIFECYCLE, payload);
```
Comment (teodorpatras) [inline src/helpers/cards.ts]: could these all `triggerWebhook`calls be better organised in a loop?

You could define some kind of array: 

```
const params = [
  {    
      token_status: ProvisioningTokenStatus.INACTIVE,
      message_reason: "LUK_REPLENISHMENT",
  {
      token_status: ProvisioningTokenStatus.INACTIVE,
      message_reason: "DEVICE_PROVISIONING_RESULT",
  }
  // and so on
];
```

and then just loop it:

```
for (const opt of params) {
    await triggerWebhook(CardWebhookEvent.CARD_TOKEN_LIFECYCLE, {...baseData, ...opt});
}
```

File: src/helpers/cards.ts (hunk 4 above)
Comment (teodorpatras) [reply src/helpers/cards.ts]: I added pretty-quick [here](https://github.com/kontist/mock-solaris/pull/8/files) which should make it easier to take care of all these nits.

File: src/helpers/cards.ts (hunk 6 above)
Comment (fabribet) [reply src/helpers/cards.ts]: b0d94e4

File: src/helpers/cards.ts (hunk 2 above)
Comment (fabribet) [reply src/helpers/cards.ts]: b0d94e4

File: src/helpers/cards.ts (hunk 1 above)
Comment (fabribet) [reply src/helpers/cards.ts]: b0d94e4

**Review by p-janik — APPROVED** [review]
LGTM :+1: thanks for adjustments

**Review by p-janik — APPROVED** [review]

**Comment by fabribet** [comment]
> LGTM 👍 thanks for adjustments

Thanks to you for your input, I will wait @teodorpatras to confirm he's good with the changes before merging.

**Review by teodorpatras — APPROVED** [review]
🚀

**Pull request merged**